### PR TITLE
fix(chat): handle connector consent and resolve pending requests

### PR DESCRIPTION
**Pull request by HamedMP**
## Summary

Stack 1/2. UI companion: #1561. Both layers must ship to fix invisible approval stalls.

- Handle Codex `mcpServer/elicitation/request` instead of leaving native connector execution blocked behind an unanswered server request. Supports bounded primitive forms, message-only explicit consent, and user-opened HTTPS authorization links; unsupported shapes fail closed.
- Preserve structured requests through canonical Chat and add an owner-scoped input submission endpoint through the existing provider control channel. No answer values are copied into transcript text.
- Bound pending consent, cancel on expiry/shutdown, and persist request resolution. Completed native turns clear local controls without writing to an already-closing provider pipe.
- Add protocol contract coverage using the version-matched app-server schema digest and public-safe development documentation.

This addresses missing request handling, not transport latency. A separate UI projection defect is fixed in #1561. Historical connector stalls are consistent with the reproduced missing-handler path; the original raw request frames were not retained, so exact historical attribution is not claimed.

## Tests

- Passed: `bun run typecheck`, `pnpm exec tsc --noEmit -p shell/tsconfig.json`, `bun run check:patterns` (5 warnings, 0 violations), `git diff --check`.
- Passed the combined 24-file focused suite (287 tests), then 41 direct connector boundary tests and 16 web recovery/hook/wiring tests after review updates. Includes real runner/control-socket tests and the reproduced terminal pipe race.
- Focused coverage for the three new connector modules passes the configured gates: 100% lines/functions, 99.41% statements, 99.48% branches.
- Full local `bun run test --maxWorkers=2`: 13,650 tests passed, 21 skipped; command exited nonzero after a fork-worker exit in unchanged `tests/integrations/routes.test.ts`. The whole-file rerun also hit a 60-second database `beforeEach` timeout after 50 passing tests and lost its worker. The exact timed-out Retry-After case passes alone (2.4 seconds). Neither full local invocation is claimed green; exact-head CI remains the broad gate. No out-of-scope integration code or test timeouts were changed.
- No customer data, production state, credentials, or private incident details are included in fixtures or docs.

## Review/Monitoring

- Review this backend layer independently; #1561 supplies the dependent renderers.
- Trusted Greptile 5/5 confirmed on final head `20dfa4c4bdbaf65cb8e2ae6407fdf594d56fb9b4`; all review threads resolved with evidence. `ready-for-ci` is present; [triggered CI](https://github.com/HamedMP/matrix-os/actions/runs/34124180316) passed, including all four unit shards, E2E Tests, and CI Results. Docker build and smoke checks also passed. Not merged or deployed.
- No merge or deployment authorized. Public site documentation: FinnaAI/matrix-os-site#87 (61 tests and rendered 375/768/1440-px checks passed).
- First review found unrepresentable form constraints. Fixed with seven failing-first boundary cases; minimum lengths/selections and contradictory ranges now fail immediately. Sanitized approval descriptions also survive canonical projection.

### Source of truth

- Owner-scoped Postgres Chat/thread events persist display requests and resolution. The live runner owns native request identities and whether a response can still be accepted.
- Existing provider event delivery reconciles durable UI state with live control outcomes. No new database or persistence engine.

### Lock/transaction scope

- Canonical submission performs scoped reads, then invokes the existing thread-store/control operation; it adds no multi-write transaction or network operation inside a new lock.
- Existing thread-store serialization and request/correlation validation remain authoritative. Runner pending entries are removed synchronously before async cleanup persistence.

### Acceptable orphan states

- A disconnected client can leave an unanswered request until the five-minute expiry sweep (30-second interval); expiry never grants permission.
- Terminal turns clear stale controls. Shutdown drains pending consent; unsupported or over-cap requests receive an explicit error/cancellation.
- A persist failure cancels the pending native request. No secrets are added to history as fallback input text.

### Auth source of truth

- Existing authenticated gateway principal determines the owner. Exact owner/chat/run and pending request are checked before provider dispatch; native thread correlation and question identities are validated again.
- Input route has bounded identifiers, Zod payload validation, and a 40-KiB body limit. No public unauthenticated path or automatic connector permission grant.

### Deferred scope

- Persistent connector grants, arbitrary nested JSON Schema forms, automatic browser navigation, transport migration, production rollout, and live account authorization are not included.
- Backend deployment alone does not fix missing UI controls; release with #1561 and update affected clients.
- Pre-existing Hermes native clarification responses remain tracked in #1562; this stack does not invent an unverified native reply protocol for another provider.

### CODE REVIEW

**Comment by HamedMP** [comment]
* **#1561** <a href="https://app.graphite.com/github/pr/HamedMP/matrix-os/1561?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1560** <a href="https://app.graphite.com/github/pr/HamedMP/matrix-os/1560?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/HamedMP/matrix-os/1560?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/gateway/src/coding-agents/codex-elicitation.mjs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+const Choices = z.array(z.string().max(700)).min(1).max(10);
+const TitledChoices = z.array(z.object({ const: z.string().max(700), title: Text }).strict()).min(1).max(10);
+const Field = z.object({
+  type: z.enum(["string", "number", "integer", "boolean", "array"]),
+  title: Text.nullish(), description: Text.nullish(), default: z.unknown().optional(),
+  enum: Choices.optional(), enumNames: z.array(Text).max(10).nullish(), oneOf: TitledChoices.optional(),
+  items: z.object({ type: z.literal("string").optional(), enum: Choices.optional(), anyOf: TitledChoices.optional() }).strict().optional(),
+  minimum: z.number().finite().nullish(), maximum: z.number().finite().nullish(),
+  minLength: z.number().int().nonnegative().nullish(), maxLength: z.number().int().nonnegative().nullish(),
+  minItems: z.number().int().nonnegative().nullish(), maxItems: z.number().int().nonnegative().nullish(),
```
[the remark this answers is not in the document]
Comment (HamedMP) [reply packages/gateway/src/coding-agents/codex-elicitation.mjs:29]: Fixed in f898904ae8. The compiler now rejects minimum lengths above the 400-character transport cap, selections above four or available values, zero-length/count maxima, reversed ranges, and impossible integer intervals before persisting a request. Seven failing-first cases cover the reported boundaries; compiler/provider suite passes 44 tests and the restacked 24-file suite passes 287.

**Comment by HamedMP** [comment]
@greptile-apps please re-review latest head f898904ae8b2b1a89284bb52997fc62b24bd8051. The reported impossible-form constraints are fixed with boundary regressions. All prior threads have exact fix/test replies. Current combined stack validation: 287 focused tests plus 12 web-state/wiring tests; workspace and shell typechecks and pattern scan pass.

**Comment by HamedMP** [comment]
@greptile-apps please review final head 20dfa4c4bdbaf65cb8e2ae6407fdf594d56fb9b4. Production code is unchanged from the 5/5-reviewed f898904ae8; the only addition is direct failure-boundary tests. New connector modules now pass coverage gates: 100% lines/functions, 99.41% statements, 99.48% branches. Please attach the final review to this current head.
